// loopback-next/packages/context/lib/src/index.js
"use strict";
// Copyright IBM Corp. 2017. All Rights Reserved.
// Node module: loopback
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
Object.defineProperty(exports, "__esModule", { value: true });
var binding_1 = require("./binding");
exports.Binding = binding_1.Binding;
var context_1 = require("./context");
exports.Context = context_1.Context;
var inject_1 = require("./inject");
exports.inject = inject_1.inject;
var reflect_1 = require("./reflect");
exports.NamespacedReflect = reflect_1.NamespacedReflect;
var is_promise_1 = require("./is-promise");
exports.isPromise = is_promise_1.isPromise;
// internals for testing
var resolver_1 = require("./resolver");
exports.instantiateClass = resolver_1.instantiateClass;
var inject_2 = require("./inject");
exports.describeInjectedArguments = inject_2.describeInjectedArguments;
exports.describeInjectedProperties = inject_2.describeInjectedProperties;
var reflect_2 = require("./reflect");
exports.Reflector = reflect_2.Reflector;
