// loopback-next/packages/openapi-spec-builder/lib/openapi-spec-builder.js
"use strict";
// Copyright IBM Corp. 2013,2017. All Rights Reserved.
// Node module: loopback
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
Object.defineProperty(exports, "__esModule", { value: true });
function givenOpenApiSpec(basePath) {
    return new OpenApiSpecBuilder(basePath);
}
exports.givenOpenApiSpec = givenOpenApiSpec;
class OpenApiSpecBuilder {
    constructor(basePath = '/') {
        this._spec = {
            basePath,
            paths: {},
        };
    }
    withOperation(verb, path, spec) {
        if (!this._spec.paths[path])
            this._spec.paths[path] = {};
        this._spec.paths[path][verb] = spec;
        return this;
    }
    withOperationReturningString(verb, path, operationName) {
        return this.withOperation(verb, path, {
            'x-operation-name': operationName,
            responses: {
                '200': { type: 'string' },
            },
        });
    }
    build() {
        // TODO(bajtos): deep-clone
        return this._spec;
    }
}
exports.OpenApiSpecBuilder = OpenApiSpecBuilder;
