// loopback-next/packages/authentication/lib/strategy-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@loopback/core");
/**
 * Shimmed Request to satisfy express requirements of passport strategies.
 */
class ShimRequest {
    constructor(request) {
        this.headers = request.headers;
        this.query = request.query;
        this.url = request.url;
        this.path = request.path;
        this.method = request.method;
    }
}
exports.ShimRequest = ShimRequest;
/**
 * Adapter class to invoke passport-strategy
 *   1. provides express dependencies to the passport strategies
 *   2. provides shimming of requests for passport authentication
 *   3. provides lifecycle similar to express to the passport-strategy
 *   3. provides state methods to the strategy instance
 * see: https://github.com/jaredhanson/passport
 */
class StrategyAdapter {
    /**
     * @param strategy instance of a class which implements a passport-strategy;
     * @description http://passportjs.org/
     */
    constructor(strategy) {
        this.strategy = strategy;
    }
    /**
     * The function to invoke the contained passport strategy.
     *     1. Create an instance of the strategy
     *     2. add success and failure state handlers
     *     3. authenticate using the strategy
     * @param req {http.ServerRequest} The incoming request.
     */
    authenticate(req) {
        const shimReq = new ShimRequest(req);
        return new Promise((resolve, reject) => {
            // create a prototype chain of an instance of a passport strategy
            const strategy = Object.create(this.strategy);
            // add success state handler to strategy instance
            strategy.success = function (user) {
                resolve(user);
            };
            // add failure state handler to strategy instance
            strategy.fail = function (challenge) {
                reject(new core_1.HttpErrors.Unauthorized(challenge));
            };
            // add error state handler to strategy instance
            strategy.error = function (error) {
                reject(new core_1.HttpErrors.InternalServerError(error));
            };
            // authenticate
            strategy.authenticate(shimReq);
        });
    }
}
exports.StrategyAdapter = StrategyAdapter;
